tests/bluetooth/bsim/mesh/src/mesh_test.c

```diff
@@ -567,6 +567,7 @@ bool bt_mesh_test_sync(uint32_t channel_id, uint16_t wait_sec)
 	uint32_t recv_msg;
 	int wait = wait_sec * MSEC_PER_SEC;
 
+	k_sleep(K_MSEC(100));
```

Comment:
I'm uncertain that this is needed. Unless something wrong with the back channel, `bs_bc_send_msg` should always succeed. If timeout needs to be increased, this should be done in `wait_sec`.

Reply:
Two processes might be quite far from each other. To perform synchronization between them and to make them "close" to each other not worse than 100ms they need about 100ms to perform this synchronization. 
Otherwise, the time grid and timestamps between tests might be different with different runs (depending on process desynchronization). As a result scanning cycles are occasionally but discretely different with each `va.sh` run.
Processes are synchronized when getting access to emulated Phy. After that, they get deterministic behavior.
Bsim mesh tests use backchannel before access to Phy which means synchronization should be done manually. That is done in this PR.

Try to run `va.sh` with this option with and without changes in this PR and you will see the difference

```
-argsmain -stop_on_diff
```

This delay is not related to the waiting time that is handled by test scenario synchronization that is based on backchannel usage.

Reply:
P.S. bs_bc_send_msg always succeeds even without this change but the test is not deterministic.

Reply:
>Processes are synchronized when getting access to emulated Phy. After that, they get deterministic behavior.
Bsim mesh tests use backchannel before access to Phy which means synchronization should be done manually. That is done in this PR.

Just to elaborate a bit, even though what @alxelax says is correct.

Reply:
`bs_bc_send_msg` call is non-blocking meaning that the sender doesn't require receiver to call `bs_bc_receive_msg` in advance before this call and `bs_bc_send_msg` simply calls `write()` function. So adding `k_sleep` before `bs_bc_send_msg` makes no sense for me as at this point synchronization is not required. Then, the device expects the data from the other side by calling `bs_bc_is_msg_received`, which is also non-blocking and calls `k_sleep()` if the data has not arrived yet.

Try to run va.sh with this option with and without changes in this PR and you will see the difference

-argsmain -stop_on_diff

In both cases the test fails. With the change:

```
p_2G4:  Comp: Device 0, Rx 4 differs
p_2G4:  Comp: Read:"99864,4294967295,0x8E89BED6,16,2.000000,0.000000,2,0,40,16,1000000,110474,110871,-1,0,0,0,0,110474,4,0.000000,0,"
p_2G4:  Comp:      "99864,4294967295,0x8E89BED6,16,2.000000,0.000000,2,0,40,16,1000000,129822,129823,-1,0,0,0,0,129822,4,0.000000,0,"
p_2G4:  Comp:                                                                           ^
```

without the change:

```
p_2G4:  Comp: Device 1, ModemRx 0 differs
p_2G4:  Comp: Read:"110943,0,0,2.000000,16,0.000000e+00,1.000000e+00,96.989700,100.000000,100.000000,60.000000,-60.000000,NaN, NaN"
p_2G4:  Comp:      "110943,0,0,2.000000,16,0.000000e+00,1.000000e+00,96.989700,100.000000,100.000000,35.000000,-35.000000,NaN, NaN"
p_2G4:  Comp:                                                                                        ^
```

Reply:
Did you clean up results folder before running the test? I tried running 20 times, it is not reproduced anymore.

Reply:
> `bs_bc_send_msg` call is non-blocking meaning that the sender doesn't require receiver to call `bs_bc_receive_msg` in advance before this call and `bs_bc_send_msg` simply calls `write()` function. So adding `k_sleep` before `bs_bc_send_msg` makes no sense for me as at this point synchronization is not required. Then, the device expects the data from the other side by calling `bs_bc_is_msg_received`, which is also non-blocking and calls `k_sleep()` if the data has not arrived yet.

I guess you mix test scenario synchronization and emulated processes synchronization. They are different. What you wrote is the truth for test scenario, and it works for scenarios. However, point of scenario synchronization is drifted by self between two processes.

Reply:
> Did you clean up results folder before running the test? I tried running 20 times, it is not reproduced anymore.

Ran `./tests/bluetooth/bsim/mesh/compile.sh` which removes the previous build: https://github.com/zephyrproject-rtos/zephyr/blob/49fa1519dfa7a9d095898f648daf77647867d6c4/tests/bluetooth/bsim/compile.source#L52-L57

> I guess you mix test scenario synchronization and emulated processes synchronization. They are different. What you wrote is the truth for test scenario, and it works for scenarios. However, point of scenario synchronization is drifted by self between two processes.

To synchronize 2 devices the `k_sleep` is added in `wait_for_sync`.

Reply:
> So adding k_sleep before bs_bc_send_msg

There is no need to add it before the send. That is correct. It does not harm there either.
It is just needed before the receive.

> Then, the device expects the data from the other side by calling bs_bc_is_msg_received, which is also non-blocking and calls k_sleep() if the data has not arrived yet.

Exactly. As both devices are running without any synchronization yet, 3 things can happen here:
a) both devices do the send before either device does the received() call. And both get the data before doing any k_sleep()
b) device 0 does not get the data before its k_sleep() (because device 1 did not do the send yet), it k_sleeps(100ms) and gets it after. Device 1 gets its data before it k_sleeps().
c)  device 0  gets the data before its k_sleep(). But device 1 does not, so device 1 sleeps for 100ms, and gets it after.

This is the nondeterministic behavior this PR is trying to fix.

As soon as either device does the first sleep of 100ms, because the max resync. time is 100ms, it will look itself to the phy.

Reply:
> > Did you clean up results folder before running the test? I tried running 20 times, it is not reproduced anymore.
> 
> Ran `./tests/bluetooth/bsim/mesh/compile.sh` which removes the previous build:

It is the Phy simulation results/ you need to clean up.
When you run the Phy with either the `-c`/`--check` or the `-stop_on_diff` options instead of dumping the devices radio activity to disk, it compares what it would have dumped with what there is in the disk, and complains if it sees differences. (stop_of_diff just stops the simulation as soon as it sees any difference, check mode just continues complaining)

Reply:
If this is only need to sync devices for the first time after boot, then `k_sleep` should be added at the setup function, but not into this API. This is not the only place where we use the back channel. There are (and I believe there will be) more places where we use back channel and this issue will appear again. If we want to fix it, we need to fix it properly.

Even if we leave `k_sleep` as is, if the sync time is 1 second (we forgot to call `tm_set_phy_max_resync_offset`), the issue may still happen (because this is how we make devices sync over phy, right?). So the sleep time should be the same as `max_resync_offset` in time_machine.c (https://github.com/zephyrproject-rtos/zephyr/blob/0b7a335864bfb084a3758aea066e950917675ed1/boards/posix/nrf52_bsim/time_machine.c#L48).

Reply:
shouldn't this be default behavior in the bsim backchannel API? ie if we open a backchannel, hang the process until the other end has also opened it.

Reply:
> shouldn't this be default behavior in the bsim backchannel API? ie if we open a backchannel, hang the process until the other end has also opened it.

That API is too low level to assume that.